---
title: "Differential expression analysis example: Cancer Cell Line Encyclopedia"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
---

## Summary
This notebook performs differential expression analysis on the Cancer Cell Line Encylopedia. This dataset contains > 800 cancer cell lines and > 20,000 genes. 

Note that in this dataset, there is no comparison between two groups (e.g., treatment versus control). Thus, we will define a differentially expressed gene as a gene that has higher or lower expression levels than the population mean using the Z-score method. More robust metrics such as the median absolute deviation can be substituted into this pipeline.

**UPDATES: **
  * November 4, 2020: Added more stringent and robust differential expression analysis options.
  
## 1. Load libraries
First we'll load some essential packages needed to run differential expression analysis.
```{r, warning=FALSE, message=FALSE}
# Bioinformatics libraries --> unhash if you need to install these into your workspace. 
#install.packages("BiocManager")
#BiocManager::install(c("BiocManager", "edgeR", 
#                   "DESeq2", "limma", 
#                   "sva", "PCAtools",
#                   "Glimma", "bioMart"))

#install.packages("CePa", dependences=TRUE)
#BiocManager::install("Rgraphviz")
library(CePa)
library(edgeR)
library(DESeq2)
library(limma)
library(sva)
library(PCAtools)
library(Glimma)
library(biomaRt)

# Data science libraries
#install.packages(c("tidyverse", "readr", "rrcov"))
library(tidyverse)
library(readr)
library(rrcov)

# Removes all variables in current R environment
remove(list=ls())
```

## 2. Perform data preprocessing

This code block performs essential data preprocessing for the cancer cell line encyclopedia data.

### Read in raw RNASeq counts

First, let's load the table with the raw RNASeq gene counts. The data can be downloaded from the CCLE database: https://portals.broadinstitute.org/ccle/data. The specific file I downloaded has the description: "CCLE RNAseq gene expression data for 1019 cell lines (read counts)". I have the file downloaded in a separate directory, which I call in using the `read.gct` function.
```{r}
# ACLX path
readpath = "C:\\Users\\scott\\Data\\RNASeq\\CCLE\\CCLE_RNAseq_genes_counts_20180929.gct"

# Linux
#readpath = "~/Data/RNASeq/CCLE/CCLE_RNAseq_genes_counts_20180929.gct"

ccle = read.gct(readpath)
```

Note that in the raw gene count data matrix, the rows are labeled by Ensembl IDs, while the columns have the syntax CELLLINE_TISSUE, separated by an underscore. I can keep the Ensembl gene IDs, but I need to extract the tissue type, as this is a covariate of interest.

### Map genes to Entrez ID
First, since we'll be using Entrez ID in the RECON1 model, we need to map ENSEMBL to Entrez.

Let's download the human gene map using biomaRt.
```{r}
# Get map of Ensembl to Entrez
human = useMart("ensembl", dataset="hsapiens_gene_ensembl")
genes = getBM(
  attributes=c("ensembl_gene_id", "hgnc_symbol", "entrezgene_id"),
  values=rownames(ccle),
  mart=human
)
```

The code below maps the gene identifier map and original dataset by ensembl ID.
```{r}
# Map Entrez to CCLE dataset. Remove those without Entrez ID.
rownames(ccle) <- gsub("\\..*","",rownames(ccle))

# Map rows
mapped_ccle = merge(genes, ccle, by.x="ensembl_gene_id", by.y=0)
```

Because I'm mapping this to the metabolic reconstruction, I can't use genes that don't have an Entrez ID. So I'll remove those to make my dataset even smaller.
```{r}
# Remove datapoints where entrez is NA
remove_genes = is.na(mapped_ccle[, "entrezgene_id"])
ccle = mapped_ccle[!remove_genes,]
ccle = ccle[, -c(1:2)]
agg_ccle = aggregate(ccle[, -1], 
                     by=list(ccle[, 1]), 
                     FUN=median)
colnames(agg_ccle)[1] = "entrezgene_id"
ccle = agg_ccle
```

To save memory, I'll remove variables we don't need downstream.
```{r}
rm(list = c("human", "genes", 
            "mapped_ccle", "agg_ccle", 
            "remove_genes"))
```

### Aggregate and remove duplicate cell line entries.
Note that some cell lines (columns) are duplicate entries. I need to aggregate them before moving on to additional preprocessing steps. I am specifically interested in subsetting a specific tissue lineage, so I need to grab that from the dataset descriptors.
```{r}
# Get data from the ccle dataframe
tmp = colnames(ccle)
tissue = word(string=tmp, start=2, end=-1, sep="_")
ccl = word(string=tmp, start=1, sep="_")

# Store into dataframe
colnames(ccle) = as.character(ccl)
```

Now we can being aggregating by cancer cell line. The next bit of code does a little data cleaning.
```{r}
# Get entrez gene id
tmp = t(ccle)
entrezgene = tmp[1, ]
tmp = tmp[-1, ]

ccl = rownames(tmp)
rownames(tmp) = NULL

tissue = tissue[-1]
```

The next bit of code creates a dataframe with information about the Cancer Cell Lines and its corresponding tissue lineage.
```{r}
meta = data.frame("CCL"=ccl, "Tissue"=tissue)
meta = meta[!duplicated(ccl), ]
```

The code below performs mean aggregation. If you are worried about skewed data and/or robustness, median aggregation is more appropriate.
```{r}
# Aggregate by cancer cell line
agg_tmp = aggregate(tmp, by=list(ccl), FUN=mean)
agg_tmp = t(agg_tmp)
```

Finally, we'll construct the final dataframe and call it back to the `ccle` dataframe.
```{r}
# Construct the final dataframe.
agg_tmp = as.data.frame(agg_tmp)
colnames(agg_tmp) = agg_tmp[1, ]
agg_tmp = agg_tmp[-1, ]
rownames(agg_tmp) = ccle[, 1]
ccle = agg_tmp
rm(list = c("agg_tmp"))
```

### Missing value imputation

Note that there are missing values. I kep genes where the Counts Per Million (CPM) was greater than 1 in at least 1/5 of the cancer cell lines.
```{r}
tmp = DGEList(counts=data.matrix(ccle), genes=rownames(ccle))
keep = rowSums(cpm(tmp) >= 1) >= (ncol(tmp$counts)/5)
ccle = ccle[keep, ]
```

### Separate solid tumor and hematopoietic cancer cell lines

I need to separate the hematopoeitic and leukemia (HKL) cell line lineages. This is based on domain knowledge from [Rees et al., 2016](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4718762/), where they found HKLs had different drug sensitive profiles compared to solid tumor cell lines. Then I can perform differential expression analysis on both of the them separately.
```{r}
HL_idx = (meta[, 2] == "HAEMATOPOIETIC_AND_LYMPHOID_TISSUE")

HL_CCL = ccle[ , c(HL_idx)]
HL_tissue = meta[c(HL_idx), 2]

SOLID_CCL = ccle[ , !c(HL_idx)]
SOLID_tissue = meta[!c(HL_idx), 2]
```

### Remove outliers
We will first remove some outliers in both the HL and SOLID datasets using Robust PCA. It combines two ideas to find outliers:
  1. Find projections with maximal dispersion using a grid search.
  2. Estimate the covariance using an idea similar to Hubert robust regression.

First, we'll define a function to do this filtering. I used the default thresholds for RPCA for outlier detection. For more information about RPCA, refer to [Hubert et al., 2012](https://www.tandfonline.com/doi/abs/10.1198/004017004000000563)
```{r}
# Create a function that filters data using ROBPCA
removeOutliers = function(data, id){
  """
  removeOutliers uses Robust PCA to compute the orthogonal distance from PCA and the covariate matrix to determine which datapoints are outliers. Then it separates those datapoints and returns a data object of the class geneObj.

  INPUTS:
    data: a dataframe or numerical matrix of the RNASeq counts.
    id:   a chracter or numerical vector containing the gene identifiers. 

  OUTPUT:
    geneObj: A class that has the following objects embedded in it:
      data: a dataframe or numerical matrix of the RNASeq counts that are not determined to be an outlier.
      outlier_data: a dataframe or numerical matrix of the RNASeq counts that are determined to be outliers by RPCA.
      id: A character or numerical vector containing corresponding identifiers that are not outliers.
      flag: A boolean vector that shows which rows are outliers in the original dataset.
      outlier_id: A character or numerical vector containing corresponding identifiers to outliers
  """
  pca_obj = PcaHubert(data)
  filtered_data = data[pca_obj@flag, ]
  filtered_id = id[pca_obj@flag]
  outlier_data = data[!pca_obj@flag, ]
  outlier_id = id[!pca_obj@flag]
  
  setClass(Class="geneObj",
           representation(
             data="matrix",
             outlier_data="matrix",
             id="character",
             flag="logical",
             outlier_id="character"
           )
   )
   return(new("geneObj",
          data=filtered_data,
          outlier_data=outlier_data,
          id=filtered_id,
          flag=pca_obj@flag,
          outlier_id=outlier_id))
}
```

Now let's filter out the genes.
```{r}
# Construct the input for PcaHubert of class matrix
HL_data = data.matrix(HL_CCL)
SOLID_data = data.matrix(SOLID_CCL)

# Use removeOutliers to remove noisy genes
HL_geneFilter = removeOutliers(HL_data, rownames(HL_data))
SOLID_geneFilter = removeOutliers(SOLID_data, rownames(SOLID_data))
```

Then let's filter out the tissues.
```{r}
# Now set the stage to remove noisy cancer cell lines
HL_data2 = data.matrix(HL_geneFilter@data)
SOLID_data2 = data.matrix(SOLID_geneFilter@data)

HL_data2 = t(HL_data2)
SOLID_data2 = t(SOLID_data2)

HL_allFilter = removeOutliers(HL_data2, meta[, 1])
SOLID_allFilter = removeOutliers(SOLID_data2,  meta[, 1])
```

The final dataframe will be constructed here.
```{r}
# Construct the final dataframe
HL_final = t(data.matrix(HL_allFilter@data))
SOLID_final = t(data.matrix(SOLID_allFilter@data))
rownames(HL_final) = c(HL_geneFilter@id)
rownames(SOLID_final) = c(SOLID_geneFilter@id)
final_tissue = SOLID_tissue[SOLID_allFilter@flag]
```

If we want to do a final sanity check, here's a code block visualizes the filtered tissues from RPCA using a screeplot, biplot of the first 2 principle components, and pairplot of the first 5 principle components
```{r}
tmp = t(data.matrix(SOLID_final))
colnames(tmp) = NULL
rownames(tmp) = NULL

# Visualize cancer cell lines in 2D space
solid_pc_filtered = PCAtools::pca(tmp)
PCAtools::screeplot(solid_pc_filtered)
PCAtools::biplot(solid_pc_filtered, 
                 x="PC1", y="PC2")
PCAtools::pairsplot(solid_pc_filtered)
```

And we can do the same for the gene space.
```{r}
# Visualize genes in 2D space
tmp = data.matrix(SOLID_final)
colnames(tmp) = NULL
rownames(tmp) = NULL

solid_pc_filtered = PCAtools::pca(tmp)
PCAtools::screeplot(solid_pc_filtered)
PCAtools::biplot(solid_pc_filtered)
PCAtools::pairsplot(solid_pc_filtered)
```

### Data Normalization 
Now let's take the log2 of the counts-per-million of the RNASeq counts.
```{r}
Xfinal = SOLID_final
Xfinal = DGEList(Xfinal, 
                 genes=rownames(Xfinal), 
                 group=as.factor(final_tissue))
Xfinal = calcNormFactors(Xfinal, method="TMM")
Xfinal = cpm(Xfinal, log=TRUE)
hist(Xfinal, breaks=100)
```

## 3. Differential Expression Analysis Using The Z-Score
As I explained before, because we are not comparing two groups, and because we do not have sample replicates, it would be impossible to model the data using the negative binomial distribution without imposing assumptions that may be invalid from a statistical standpoint. Thus, the comparison I made uses the Z-score and computes the p-value from that distribution.

First, we'll compute the Z-score and its corresponding distribution from a Normal distribution.
```{r}
Xfinal = SOLID_final
mu = rowMeans(Xfinal)
sd = rowSds(Xfinal)
Xfinal = (Xfinal - mu) / sd
hist(Xfinal, breaks=100)
pvalue=2*pnorm(-abs(Xfinal))
```

Next, I will adjust the p-values using the FDR method. 
```{r}
tmp = as.vector(pvalue)
tmp = p.adjust(tmp, method="fdr")
pvalue = matrix(tmp, 
                nrow=nrow(pvalue), ncol=ncol(pvalue), 
                byrow=TRUE)
colnames(pvalue) = colnames(Xfinal)
rownames(pvalue) = rownames(Xfinal)

```
That's it!

## 4. Save the file
The code block below saves a matrix of the Z-scores and FDR adjusted P-values into an Excel sheet.
```{r}
install.packages("openxlsx")
library(openxlsx)

df_list = list("Z-scores"=Xfinal, "P-value"=pvalue)
filename = "C:/Users/scott/Data/RNASeq/CCLE/DEG_RNASEQ_120920.xlsx"
write.xlsx(df_list, file=filename, row.names=TRUE)
```

## Summary
This notebook demonstrates how to process the CCLE dataset, where the input data was RNASeq counts from a gct file. The data was mapped to Entrez IDs and was aggregated. Robust PCA was used to identify outlier genes and cell lines in the dataset. Finally, we computed the necessary statistics required to label up- and downregulated genes. 